### ISSUE TITLE
[Feature] Adding fixes for full MATLAB compatibility

### ISSUE THREAD

**Issue by Dev-iL** [Revivius/nb-darcula#100]
As it happens, [MATLAB](http://www.mathworks.com/products/matlab/) comes bundled with `netbeans.jar`, and the MATLAB editor relies heavily (if not exclusively) on the NetBeans editor. For this reason, it makes sense (at least to me) to use a version of the L&F that is better adapted to NetBeans (...this repository) for the purpose of incorporating better MATLAB support. 

The incompatibility with MATLAB has the following aspects, also demonstrated in the images below:
- [ ] 1. The line number panel is not colored.
- [ ] 2. Editor tabs are not colored.
- [ ] 3. "Workbench" (`UIPanel`?) tabs retain the stock blue\bright-gray color even though they should be dark, which causes buttons to stick out in an extremely discordant and inharmonious manner.
- [ ] 4. Bottom status bar changes color when command window is focused, thus making it unreadable. The color should ideally remain dark as in the unfocused state.
- [ ] 5. Menubar text is dark instead of bright and is barely readable with the dark L&F background. When item is clicked (when choosing sub-items), the text becomes the 
- [ ] 6. Menu items are shown as checkboxes with text instead of "textviews".

**BTW, it might be that all of these issues are already fixed, and I'm just "using it wrong"...**

![Problems 1&2](http://i.imgur.com/DHbnVqx.png)
---------
![Problem 3](http://i.imgur.com/PNXIdlM.png)
---------
![Problem 4](http://i.imgur.com/KPnC9Fa.png)
---------
![Problems 5&6](http://i.imgur.com/do7vHJ5.png)

**Comment by Revivius** [comment]
@Dev-iL; I dont know how MATLAB and NetBeans integrates. I dont think MATLAB GUI is built on top of NetBeans platform, is it ?

Just by looking at the screenshots you attached, I dont think there is anything that can be done within this project and custom swing components used by MATLAB should take different LAFs into account. Still; out of curiosity; I have downloaded trial  version of MATLAB and have no idea how to end up with the screens you provided. Can you please explain steps ?

**Comment by Dev-iL** [comment]
@Revivius Here are the steps you need to take:

1. Inside MATLAB, run the command `prefdir` to get your preferences folder.
2. Open that folder and create a `javaclasspath.txt` file. Inside this file add the path to the darcula `.jar` (in my case it's `E:\GitHub\nb-darcula\lib\com\bulenkov\darcula\1.0.0\darcula-1.0.0.jar`). This action adds the the classes within the jar to MATLAB's static classpath, which makes them "visible" on MATLAB boot.
3. In the MATLAB command execute `edit('startup.m')`. In this file add:
```
javax.swing.UIManager.setLookAndFeel('com.bulenkov.darcula.DarculaLaf');
f = com.mathworks.mde.desk.MLDesktop.getInstance.getMainFrame();
javax.swing.SwingUtilities.updateComponentTreeUI(f);
```
4. Save `startup.m` and restart MATLAB. The new L&F should now be applied.

Thanks for your interest :)

**Comment by markiewb** [comment]
IMO we should concentrate on issues regarding NB. 

Other 3rd party LAFs will also have the reported issues. Nevertheless there can be fork for MATLAB.

**Comment by Dev-iL** [comment]
@markiewb I won't argue with you there, obviously this feature deviates from the main purpose of the repository. It's just a matter of how much...
Regarding the 3rd party LAFs - some do, some don't. I would be more than happy to fix all of the above myself if only I knew what to look for :|

**Comment by markiewb** [comment]
@Dev-iL: @Revivius can fix issues because he has access to the sources of NetBeans. Matlab is closed-source AFAIK, so it would be trail-and-error fixing issues.

Perhaps it is also worth to file an issue at MATLABs bugtracker?!

**Comment by Dev-iL** [comment]
@markiewb You are almost entirely right. MATLAB's GUI is built on Java and comes with a bunch of `jars` that define some form of an API (such as the `netbeans.jar` I mentioned initially). While it is true that some of MATLAB's internals are closed-source, the `.jar`s are viewable using a disassembler (owing to the fact they are practically not obfuscated at all). For this reason, it isn't exactly "trial and error" where there's absolutely no reference available, but it's also not fair to claim that the full source code is there. Moreover, you can inspect the UI using the tool I mentioned, and you can query various settings of the GUI (the P.S.2 command in the OP post returns some > 1100 settings).
The bottom line is - **if there's a will, there's a way**.

Submitting a feature request to MATLAB about something as obscure and undocumented (and one might add also _niche_ and _unimportant_) is doomed to fail and not even worth the trouble (things that are more important than the theme of the interface take years to fix...)

**Comment by Revivius** [comment]
@Dev-iL; I followed your steps in MATLAB can see reported problems now.

To make some points clear: this repository contains NetBeans specific 'glue code' (since NetBeans uses many custom swing components) to apply original Darcula LAF. It is true that we are now using forked version of original Darcula LAF and some of the problems you reported can be solved by modifying original Darcula LAF but most of them (eg tabs from screenshot #1 ) are related to how MATLAB custom swing components were written.

So, I agree with @markiewb that we should concentrate on NB integration within this repository and you should report this to MATLAB.

**Comment by Dev-iL** [comment]
@Revivius Ok then. Feel free to tag this as [wontfix] and close the issue.

**Comment by Dev-iL** [comment]
Below is one of the classes from `mwswing.jar`(© _The MathWorks, Inc._ ):
Do you guys think that extending it and adding a "correctDarculaLookAndFeel()" with some appropriate settings will solve the problems?

```
package com.mathworks.mwswing.plaf;

import com.jgoodies.looks.Options;
import com.jgoodies.looks.plastic.Plastic3DLookAndFeel;
import com.jgoodies.looks.plastic.PlasticTheme;
import com.mathworks.mwswing.ColorUtils;
import com.mathworks.mwswing.FontUtils;
import com.mathworks.mwswing.MJUtilities;
import com.mathworks.mwswing.MacAppearanceUtils;
import com.mathworks.util.LanguageUtils;
import com.mathworks.util.PlatformInfo;
import com.sun.java.swing.plaf.windows.WindowsTreeUI.CollapsedIcon;
import com.sun.java.swing.plaf.windows.WindowsTreeUI.ExpandedIcon;
import java.awt.Color;
import java.awt.Component;
import java.awt.Font;
import java.awt.Graphics;
import java.awt.GraphicsEnvironment;
import java.awt.SystemColor;
import java.awt.Toolkit;
import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.io.PrintStream;
import javax.swing.BorderFactory;
import javax.swing.Icon;
import javax.swing.UIDefaults;
import javax.swing.UIManager;
import javax.swing.UnsupportedLookAndFeelException;
import javax.swing.border.CompoundBorder;

public class PlafUtils
{
  private static boolean sLookAndFeelAlreadySet = false;
  private static final int SIZE = 9;
  private static final int HALF_SIZE = 4;
  
  public static boolean isMotifLookAndFeel()
  {
    return UIManager.getLookAndFeel().getClass().getName().equals("com.sun.java.swing.plaf.motif.MotifLookAndFeel");
  }
  
  public static boolean isPlasticLookAndFeel()
  {
    return UIManager.getLookAndFeel().getClass().getName().equals("com.jgoodies.looks.plastic.Plastic3DLookAndFeel");
  }
  
  public static boolean isWindowsLookAndFeel()
  {
    return UIManager.getLookAndFeel().getClass().getName().equals("com.sun.java.swing.plaf.windows.WindowsLookAndFeel");
  }
  
  public static boolean isMetalLookAndFeel()
  {
    return UIManager.getLookAndFeel().getClass().getName().equals("javax.swing.plaf.metal.MetalLookAndFeel");
  }
  
  public static boolean isAquaLookAndFeel()
  {
    return PlatformInfo.isMacintosh();
  }
  
  public static void correctPlatformLookAndFeel()
  {
    if (isAquaLookAndFeel()) {
      correctMacintoshLookAndFeel();
    } else if (isWindowsLookAndFeel()) {
      correctWindowsLookAndFeel();
    } else if (isPlasticLookAndFeel()) {
      correctPlasticLookAndFeel();
    }
  }
  
  public static void correctWindowsLookAndFeel()
  {
    correctWindowsCaretBlinkRate();
    correctWindowsFonts();
    trackWindowsSystemColors();
    updateTreeIcons();
    
    UIManager.put("TextField.disabledBackground", UIManager.getColor("control"));
    UIManager.put("TextField.inactiveBackground", UIManager.getColor("control"));
    
    UIManager.put("PopupMenu.consumeEventOnClose", Boolean.valueOf(false));
    if (PlatformInfo.isWindowsModernAppearance()) {
      correctXPMenuBorders();
    }
    if (MJUtilities.isHighContrast()) {
      UIManager.put("Table.scrollPaneBorder", BorderFactory.createLineBorder((Color)UIManager.get("windowBorder"), 1));
    }
  }
  
  public static void correctPlasticLookAndFeel()
  {
    UIManager.put("MenuItem.acceleratorDelimiter", "+");
    
    UIManager.put("Button.defaultButtonFollowsFocus", Boolean.TRUE);
    
    updateTreeIcons();
    
    UIManager.put("CheckBox.icon", new CorrectedPlasticCheckBoxIcon(UIManager.getIcon("CheckBox.icon"), null));
  }
  
  public static void correctMacintoshLookAndFeel()
  {
    MacAppearanceUtils.initialize();
    
    System.setProperty("apple.laf.useScreenMenuBar", "false");
    
    System.setProperty("apple.awt.showGrowBox", "true");
  }
  
  public static void correctWindowsCaretBlinkRate()
  {
    Integer localInteger = Integer.valueOf(MJUtilities.getCaretBlinkRate());
    UIManager.put("TextField.caretBlinkRate", localInteger);
    UIManager.put("FormattedTextField.caretBlinkRate", localInteger);
    UIManager.put("PasswordField.caretBlinkRate", localInteger);
    UIManager.put("TextArea.caretBlinkRate", localInteger);
    UIManager.put("TextPane.caretBlinkRate", localInteger);
    UIManager.put("EditorPane.caretBlinkRate", localInteger);
  }
  
  public static void correctWindowsFonts()
  {
    Font localFont = FontUtils.getSystemUIFont();
    if (LanguageUtils.isCJK())
    {
      UIManager.put("Menu.font", localFont);
      UIManager.put("MenuItem.font", localFont);
      UIManager.put("CheckBoxMenuItem.font", localFont);
      UIManager.put("RadioButtonMenuItem.font", localFont);
    }
    UIManager.getDefaults().put("Button.font", localFont);
    UIManager.getDefaults().put("CheckBox.font", localFont);
    UIManager.getDefaults().put("ComboBox.font", localFont);
    UIManager.getDefaults().put("EditorPane.font", localFont);
    UIManager.getDefaults().put("FormattedTextField.font", localFont);
    UIManager.getDefaults().put("Label.font", localFont);
    UIManager.getDefaults().put("TextArea.font", localFont);
    UIManager.getDefaults().put("List.font", localFont);
    UIManager.getDefaults().put("Panel.font", localFont);
    UIManager.getDefaults().put("ProgressBar.font", localFont);
    UIManager.getDefaults().put("RadioButton.font", localFont);
    UIManager.getDefaults().put("ScrollPane.font", localFont);
    UIManager.getDefaults().put("TabbedPane.font", localFont);
    UIManager.getDefaults().put("Table.font", localFont);
    UIManager.getDefaults().put("TableHeader.font", localFont);
    UIManager.getDefaults().put("TextField.font", localFont);
    UIManager.getDefaults().put("TextPane.font", localFont);
    UIManager.getDefaults().put("TitledBorder.font", localFont);
    UIManager.getDefaults().put("ToggleButton.font", localFont);
    UIManager.getDefaults().put("Tree.font", localFont);
    UIManager.getDefaults().put("Viewport.font", localFont);
  }
  
  public static void setLookAndFeel()
  {
    if (!sLookAndFeelAlreadySet)
    {
      try
      {
        if (PlatformInfo.isXWindows()) {
          setUnixLookAndFeel();
        } else {
          UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
        }
      }
      catch (UnsupportedLookAndFeelException localUnsupportedLookAndFeelException)
      {
        System.out.println("Look & Feel not supported.");
      }
      catch (IllegalAccessException localIllegalAccessException)
      {
        System.out.println("Look & Feel could not be accessed.");
      }
      catch (ClassNotFoundException localClassNotFoundException)
      {
        System.out.println("Look & Feel could not be found.");
      }
      catch (InstantiationException localInstantiationException)
      {
        System.out.println("Look & Feel could not be instantiated.");
      }
      correctPlatformLookAndFeel();
      sLookAndFeelAlreadySet = true;
    }
  }
  
  private static void setUnixLookAndFeel()
    throws ClassNotFoundException, InstantiationException, IllegalAccessException, UnsupportedLookAndFeelException
  {
    if (GraphicsEnvironment.isHeadless()) {
      return;
    }
    Options.setUseNarrowButtons(false);
    
    Plastic3DLookAndFeel.setPlasticTheme((PlasticTheme)Class.forName("com.jgoodies.looks.plastic.theme.ExperienceRoyale").newInstance());
    
    UIManager.setLookAndFeel("com.jgoodies.looks.plastic.Plastic3DLookAndFeel");
  }
  
  private static class CorrectedPlasticCheckBoxIcon
    implements Icon
  {
    private final Icon fIcon;
    
    private CorrectedPlasticCheckBoxIcon(Icon paramIcon)
    {
      this.fIcon = paramIcon;
    }
    
    public void paintIcon(Component paramComponent, Graphics paramGraphics, int paramInt1, int paramInt2)
    {
      paramGraphics.setColor(Plastic3DLookAndFeel.getControl());
      
      paramGraphics.fillRect(paramInt1, paramInt2, 13, 13);
      this.fIcon.paintIcon(paramComponent, paramGraphics, paramInt1, paramInt2);
    }
    
    public int getIconWidth()
    {
      return this.fIcon.getIconWidth();
    }
    
    public int getIconHeight()
    {
      return this.fIcon.getIconHeight();
    }
  }
  
  private static void correctXPMenuBorders()
  {
    CompoundBorder localCompoundBorder = BorderFactory.createCompoundBorder(BorderFactory.createMatteBorder(1, 1, 1, 1, UIManager.getColor("controlShadow")), BorderFactory.createEmptyBorder(2, 2, 2, 2));
    
    UIManager.put("PopupMenu.border", localCompoundBorder);
  }
  
  private static void trackWindowsSystemColors()
  {
    Toolkit localToolkit = Toolkit.getDefaultToolkit();
    SystemColorTracker localSystemColorTracker = new SystemColorTracker(null);
    localToolkit.addPropertyChangeListener("win.3d.backgroundColor", localSystemColorTracker);
    localToolkit.addPropertyChangeListener("win.3d.highlightColor", localSystemColorTracker);
    localToolkit.addPropertyChangeListener("win.3d.lightColor", localSystemColorTracker);
    localToolkit.addPropertyChangeListener("win.3d.shadowColor", localSystemColorTracker);
  }
  
  private static class SystemColorTracker
    implements PropertyChangeListener
  {
    public void propertyChange(PropertyChangeEvent paramPropertyChangeEvent)
    {
      String str = paramPropertyChangeEvent.getPropertyName();
      Color localColor;
      SystemColor localSystemColor;
      if (str.equals("win.3d.backgroundColor"))
      {
        localColor = UIManager.getColor("control");
        localSystemColor = SystemColor.control;
        if (localColor.getRGB() != localSystemColor.getRGB()) {
          UIManager.put("control", localSystemColor);
        }
        localColor = UIManager.getColor("textText");
        localSystemColor = SystemColor.textText;
        if (localColor.getRGB() != localSystemColor.getRGB()) {
          UIManager.put("textText", localSystemColor);
        }
        localColor = UIManager.getColor("textInactiveText");
        localSystemColor = SystemColor.textInactiveText;
        if (localColor.getRGB() != localSystemColor.getRGB()) {
          UIManager.put("textInactiveText", localSystemColor);
        }
      }
      else if (str.equals("win.3d.highlightColor"))
      {
        localColor = UIManager.getColor("controlLtHighlight");
        localSystemColor = SystemColor.controlLtHighlight;
        if (localColor.getRGB() != localSystemColor.getRGB()) {
          UIManager.put("controlLtHighlight", localSystemColor);
        }
      }
      else if (str.equals("win.3d.lightColor"))
      {
        localColor = UIManager.getColor("controlHighlight");
        localSystemColor = SystemColor.controlHighlight;
        if (localColor.getRGB() != localSystemColor.getRGB()) {
          UIManager.put("controlHighlight", localSystemColor);
        }
      }
      else if (str.equals("win.3d.shadowColor"))
      {
        localColor = UIManager.getColor("controlShadow");
        localSystemColor = SystemColor.controlShadow;
        if (localColor.getRGB() != localSystemColor.getRGB()) {
          UIManager.put("controlShadow", localSystemColor);
        }
        localColor = UIManager.getColor("controlDkShadow");
        localSystemColor = SystemColor.controlDkShadow;
        if (localColor.getRGB() != localSystemColor.getRGB()) {
          UIManager.put("controlDkShadow", localSystemColor);
        }
      }
    }
  }
  
  public static void updateTreeIcons()
  {
    UIManager.put("Tree.expandedIcon", new ContrastingExpandedIcon(null));
    UIManager.put("Tree.collapsedIcon", new ContrastingCollapsedIcon(null));
  }
  
  private static class ContrastingExpandedIcon
    extends WindowsTreeUI.ExpandedIcon
  {
    public void paintIcon(Component paramComponent, Graphics paramGraphics, int paramInt1, int paramInt2)
    {
      if ((PlatformInfo.isWindowsModernAppearance()) || (!ColorUtils.isDark(paramComponent.getBackground())))
      {
        super.paintIcon(paramComponent, paramGraphics, paramInt1, paramInt2);
      }
      else
      {
        paramGraphics.setColor(paramComponent.getBackground());
        paramGraphics.fillRect(paramInt1, paramInt2, 8, 8);
        paramGraphics.setColor(Color.gray);
        paramGraphics.drawRect(paramInt1, paramInt2, 8, 8);
        paramGraphics.setColor(Color.white);
        paramGraphics.drawLine(paramInt1 + 2, paramInt2 + 4, paramInt1 + 6, paramInt2 + 4);
      }
    }
  }
  
  private static class ContrastingCollapsedIcon
    extends WindowsTreeUI.CollapsedIcon
  {
    public void paintIcon(Component paramComponent, Graphics paramGraphics, int paramInt1, int paramInt2)
    {
      if ((PlatformInfo.isWindowsModernAppearance()) || (!ColorUtils.isDark(paramComponent.getBackground())))
      {
        super.paintIcon(paramComponent, paramGraphics, paramInt1, paramInt2);
      }
      else
      {
        paramGraphics.setColor(paramComponent.getBackground());
        paramGraphics.fillRect(paramInt1, paramInt2, 8, 8);
        paramGraphics.setColor(Color.gray);
        paramGraphics.drawRect(paramInt1, paramInt2, 8, 8);
        paramGraphics.setColor(Color.white);
        paramGraphics.drawLine(paramInt1 + 2, paramInt2 + 4, paramInt1 + 6, paramInt2 + 4);
        paramGraphics.drawLine(paramInt1 + 4, paramInt2 + 2, paramInt1 + 4, paramInt2 + 6);
      }
    }
  }
}
```

**Comment by Revivius** [comment]
@Dev-i; just by looking at the code; this seems to be a promising place to start with (assuming Swing Components that cause inconsistencies with the LAF were written to lookup colors and borders to use from UIManager). But once again, I have no idea how MATLAB sets up the LAF and how custom components were written.

I can understand that you like the LAF and want to use it with MATLAB but this repo is really not the correct place to discuss. Closing the issue.
